rust/operator/src/lib.rs

```diff
@@ -0,0 +1,672 @@
+
+                    let state = sticky_scheduler.schedule(
+                        &pod_id_factory,
+                        &RoleGroupEligibleNodes::from(&self.eligible_nodes),
+                        &self.existing_pods,
+                    )?;
+
+                    let mapping = state.remaining_mapping().filter(
+                        APP_NAME,
+                        &self.context.name(),
+                        role_str,
+                        role_group,
+                    );
+
+                    if let Some((pod_id, node_id)) = mapping.iter().next() {
+                        // now we have a node that needs a pod -> get validated config
+                        let validated_config = config_for_role_and_group(
+                            pod_id.role(),
+                            pod_id.group(),
+                            &self.validated_role_config,
+                        )?;
+
+                        let config_maps = self.create_config_maps(pod_id, validated_config).await?;
+
+                        self.create_pod(
+                            pod_id,
+                            &role,
+                            &node_id.name,
+                            &config_maps,
+                            validated_config,
+                        )
+                        .await?;
+
+                        history.save(&self.context.resource).await?;
+
+                        return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+                    }
+                }
+            }
+        }
+
+        // If we reach here it means all pods must be running on target_version.
+        // We can now set current_version to target_version (if target_version was set) and
+        // target_version to None
+        finalize_versioning(&self.context.client, &self.context.resource).await?;
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    /// Creates the config maps required for a hive instance (or role, role_group combination):
+    /// * The 'hive-site.xml' properties file
+    /// * The 'log4j.properties'
+    ///
+    /// Returns a map with a 'type' identifier (e.g. data, id) as key and the corresponding
+    /// ConfigMap as value. This is required to set the volume mounts in the pod later on.
+    ///
+    /// # Arguments
+    ///
+    /// - `pod_id` - The `PodIdentity` containing app, instance, role, group names and the id.
+    /// - `validated_config` - The validated product config.
+    /// - `id_mapping` - All id to node mappings required to create config maps
+    ///
+    async fn create_config_maps(
+        &self,
+        pod_id: &PodIdentity,
+        validated_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+    ) -> Result<HashMap<&'static str, ConfigMap>, Error> {
+        let mut config_maps = HashMap::new();
+        let mut config_maps_data = BTreeMap::new();
+
+        let log4j_config = include_str!("log4j.properties");
```

Comment:
This was just something that I initially implemented to get this of the ground quickly, not sure if we should merge this tbh.
How are we treating log4j files for the other products?

Reply:
This only came up with the hdfs and hive. What would you suggest?